micropython/lora/lora/async_modem.py

```diff
@@ -0,0 +1,42 @@
+# MicroPython LoRa async modem driver
+# MIT license; Copyright (c) 2022 Angus Gratton
+#
+# LoRa is a registered trademark or service mark of Semtech Corporation or its affiliates.
+
+import uasyncio
+
+
+class AsyncModem:
+    def __init__(self, modem):
+        self._modem = modem
+        self._flag = uasyncio.ThreadSafeFlag()
+        self._modem.set_irq_callback(self._callback)
+
+    def configure(self, lora_cfg, rx_calibrate=False):
+        return self._modem.configure(lora_cfg, rx_calibrate)
+
+    async def receive(self, timeout_ms=None):
+        # TODO: check for any pending TX
+        self._flag.clear()
+        will_cb = self._modem.start_receive(timeout_ms)
+        if will_cb:
+            await self._flag.wait()
+            assert self._modem.is_receive_completed()
+        else:
+            while not self._modem.is_receive_completed():
+                await uasyncio.sleep_ms(0)
+        return self._modem.end_receive(timeout_ms)
+
+    async def transmit(self, message):
+        self._flag.clear()
+        will_cb = self._modem.start_transmit(message)
+        if will_cb:
+            await self._flag.wait()
+            assert self._modem.is_transmit_completed()
+        else:
+            while not self._modem.is_transmit_completed():
+                await uasyncio.sleep_ms(0)
+        self._modem.end_transmit()
+
+    def _callback(self):
+        self._flag.set()
```

Comment:
What happens if I first call `receive` and then while it's waiting I call `transmit`?

Reply:
Yes, this is not at all supported right now. Need to either make it clear that the same task should be calling both receive and transmit, or re-add support for cancelling receive.

Reply:
Now supported!